apps/extension/client/src/checkTypegenNesting.ts

```diff
@@ -0,0 +1,166 @@
+
+const typegenPatternKeys = ["*.ts", "*.tsx", "*.mts", "*.cts"];
+const typegenPattern = "${capture}.typegen.ts";
+
+const getXStateConfig = () => vscode.workspace.getConfiguration("xstate");
+const getFileNestingConfig = () =>
+  vscode.workspace.getConfiguration("explorer.fileNesting");
+
+// Checks if the user has our typegen pattern set in their workspace config
+const hasTypeGenPattern = (fileNestingPatterns: object) =>
+  Object.values(fileNestingPatterns).some((pattern: string) =>
+    pattern
+      .split(",")
+      .map((part) => part.trim())
+      .some((part) => part === typegenPattern)
+  );
+
+const createTypeGenPatterns = (transformer: (patternKey: string) => string) => {
+  const transformedArray = typegenPatternKeys.map((patternKey) => ({
+    [patternKey]: transformer(patternKey),
+  }));
+  // Return an object with all the transformed keys
+  return transformedArray.reduce(function (result, item) {
+    var key = Object.keys(item)[0]; //first property: a, b, c
+    result[key] = item[key];
+    return result;
+  }, {});
+};
+
+const enableTypegenNesting = () => {
+  const fileNestingConfig = getFileNestingConfig();
+
+  const fileNestingPatterns = fileNestingConfig.get("patterns");
+  if (typeof fileNestingPatterns !== "object") return;
+
+  // If the user chooses to enable file nesting for typegen files we need to make sure that VSCode's file nesting is also enabled
+  fileNestingConfig.update("enabled", true, true);
+
+  const getUpdatedPattern = (typegenPatternKey: string) => {
+    // Check if the user has a pattern defined for ts files
+    const existingTsPattern = fileNestingPatterns[typegenPatternKey];
+
+    // If the user has defined a pattern for ts files we add our typegen pattern to it, otherwise we create a new pattern
+    return existingTsPattern
+      ? `${existingTsPattern}, ${typegenPattern}`
+      : typegenPattern;
+  };
+
+  const updatedPatterns = createTypeGenPatterns(getUpdatedPattern);
+
+  // Update file nesting patterns with all existing patterns and our updated pattern
+  fileNestingConfig.update(
+    "patterns",
+    {
+      ...fileNestingPatterns, // Reuse existing patterns
+      ...updatedPatterns, // Add our updated patterns
+    },
+    true
+  );
+};
+
+const removeEmptyPatterns = (obj: object) =>
+  Object.fromEntries(
+    Object.entries(obj).filter(([_, pattern]) => pattern != "")
+  );
+
+const disableTypegenNesting = () => {
+  const fileNestingConfig = getFileNestingConfig();
+
+  const fileNestingPatterns = fileNestingConfig.get("patterns");
+  if (typeof fileNestingPatterns !== "object") return;
+
+  const getUpdatedPattern = (typegenPatternKey: string) => {
+    // Check if the user has a pattern defined for ts files
+    const existingTsPattern: string | undefined =
+      fileNestingPatterns[typegenPatternKey];
+
+    // If there's no patterns defined for ts files we don't need to do anything
+    if (!existingTsPattern || existingTsPattern === "") return undefined;
```

Comment:
If I read the code correctly then according to the declared signature of `createTypeGenPatterns` this should be a type error because the callback type there is typed as `(patternKey: string) => string`